File: ViewModels/MainViewModel.cs
using System.Collections.ObjectModel;
using BreedFinder.Models;
using BreedFinder.Services;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace BreedFinder.ViewModels;

public partial class MainViewModel : ObservableObject
{
    private readonly IBreedDataService _breedDataService;
    private readonly IDogImageService _imageService;

    // ── Slider fields (0–10) ──────────────────────────────────────────────────

    [ObservableProperty] private double sizeValue = 5;
    [ObservableProperty] private double energyLevelValue = 5;
    [ObservableProperty] private double sheddingValue = 5;
    [ObservableProperty] private double guardednessValue = 5;
    [ObservableProperty] private double aggressivenessValue = 5;
    [ObservableProperty] private double immunityValue = 5;
    [ObservableProperty] private double lifespanValue = 5;
    [ObservableProperty] private double groomingNeedsValue = 5;
    [ObservableProperty] private double vetVisitsValue = 5;
    [ObservableProperty] private double loyaltyValue = 5;
    [ObservableProperty] private double singleOwnerValue = 5;

    // ── Slider labels ─────────────────────────────────────────────────────────

    public string SizeLabel => GetSizeLabel(SizeValue);
    public string EnergyLevelLabel => GetEnergyLabel(EnergyLevelValue);
    public string SheddingLabel => GetSheddingLabel(SheddingValue);
    public string GuardednessLabel => GetGuardednessLabel(GuardednessValue);
    public string AggressivenessLabel => GetAggressivenessLabel(AggressivenessValue);
    public string ImmunityLabel => GetImmunityLabel(ImmunityValue);
    public string LifespanLabel => GetLifespanLabel(LifespanValue);
    public string GroomingNeedsLabel => GetGroomingLabel(GroomingNeedsValue);
    public string VetVisitsLabel => GetVetVisitsLabel(VetVisitsValue);
    public string LoyaltyLabel => GetLoyaltyLabel(LoyaltyValue);
    public string SingleOwnerLabel => GetSingleOwnerLabel(SingleOwnerValue);

    partial void OnSizeValueChanged(double value) => OnPropertyChanged(nameof(SizeLabel));
    partial void OnEnergyLevelValueChanged(double value) => OnPropertyChanged(nameof(EnergyLevelLabel));
    partial void OnSheddingValueChanged(double value) => OnPropertyChanged(nameof(SheddingLabel));
    partial void OnGuardednessValueChanged(double value) => OnPropertyChanged(nameof(GuardednessLabel));
    partial void OnAggressivenessValueChanged(double value) => OnPropertyChanged(nameof(AggressivenessLabel));
    partial void OnImmunityValueChanged(double value) => OnPropertyChanged(nameof(ImmunityLabel));
    partial void OnLifespanValueChanged(double value) => OnPropertyChanged(nameof(LifespanLabel));
    partial void OnGroomingNeedsValueChanged(double value) => OnPropertyChanged(nameof(GroomingNeedsLabel));
    partial void OnVetVisitsValueChanged(double value) => OnPropertyChanged(nameof(VetVisitsLabel));
    partial void OnLoyaltyValueChanged(double value) => OnPropertyChanged(nameof(LoyaltyLabel));
    partial void OnSingleOwnerValueChanged(double value) => OnPropertyChanged(nameof(SingleOwnerLabel));

    // ── Results state ─────────────────────────────────────────────────────────

    [ObservableProperty] private ObservableCollection<DogBreed> matchingBreeds = [];
    [ObservableProperty] private bool filtersExpanded = true;
    [ObservableProperty] private bool isResultsVisible;
    [ObservableProperty] private bool isLoading;
    [ObservableProperty] private string resultsHeader = string.Empty;
    [ObservableProperty] private DogBreed? selectedBreed;

    public bool IsNotLoading => !IsLoading;
    partial void OnIsLoadingChanged(bool value) => OnPropertyChanged(nameof(IsNotLoading));

    public MainViewModel(IBreedDataService breedDataService, IDogImageService imageService)
    {
        _breedDataService = breedDataService;
        _imageService = imageService;

        // Pre-warm the image cache in the background on app start
        _imageService.PrewarmAsync(_breedDataService.GetAllBreeds()
            .Select(b => b.DogCeoBreedKey)
            .Where(k => !string.IsNullOrEmpty(k)));
    }

    // ── Navigation ────────────────────────────────────────────────────────────

    partial void OnSelectedBreedChanged(DogBreed? value)
    {
        if (value is null) return;
        Shell.Current.GoToAsync("BreedDetailPage",
            new Dictionary<string, object> { ["Breed"] = value });
        SelectedBreed = null;
    }

    // ── Commands ──────────────────────────────────────────────────────────────

    [RelayCommand]
    private async Task FindBreedsAsync()
    {
        IsLoading = true;
        IsResultsVisible = false;
        FiltersExpanded = false;

        // Matching is CPU-bound; run off the UI thread
        var results = await Task.Run(() => _breedDataService.GetMatchingBreeds(
            SizeValue, EnergyLevelValue, SheddingValue,
            GuardednessValue, AggressivenessValue, ImmunityValue,
            LifespanValue, GroomingNeedsValue, VetVisitsValue,
            LoyaltyValue, SingleOwnerValue));

        MatchingBreeds.Clear();
        foreach (var b in results) MatchingBreeds.Add(b);

        ResultsHeader = $"{results.Count} breeds found — sorted by best match";
        IsLoading = false;
        IsResultsVisible = true;

        // Fetch images in the background; each breed's ImageUrl property fires
        // PropertyChanged so individual cards update as images arrive
        _ = FetchImagesAsync(results);
    }

    [RelayCommand]
    private void ResetFilters()
    {
        SizeValue = 5; EnergyLevelValue = 5; SheddingValue = 5;
        GuardednessValue = 5; AggressivenessValue = 5; ImmunityValue = 5;
        LifespanValue = 5; GroomingNeedsValue = 5; VetVisitsValue = 5;
        LoyaltyValue = 5; SingleOwnerValue = 5;
    }

    [RelayCommand]
    private void ToggleFilters() => FiltersExpanded = !FiltersExpanded;

    // ── Background image fetching ─────────────────────────────────────────────

    private async Task FetchImagesAsync(IReadOnlyList<DogBreed> breeds)
    {
        using var semaphore = new SemaphoreSlim(4, 4);
        var tasks = breeds.Select(async breed =>
        {
            await semaphore.WaitAsync();
            try
            {
                var url = await _imageService.GetImageUrlAsync(breed.DogCeoBreedKey);
                if (!string.IsNullOrEmpty(url))
                    breed.ImageUrl = url;
            }
            finally { semaphore.Release(); }
        });
        await Task.WhenAll(tasks);
    }

    // ── Label helpers ─────────────────────────────────────────────────────────

    private static string GetSizeLabel(double v) => v switch
    {
        <= 2 => "Tiny", <= 4 => "Small", <= 6 => "Medium", <= 8 => "Large", _ => "Giant"
    };
    private static string GetEnergyLabel(double v) => v switch
    {
        <= 2 => "Couch Potato", <= 4 => "Low Energy", <= 6 => "Moderate",
        <= 8 => "High Energy", _ => "Hyperactive"
    };
    private static string GetSheddingLabel(double v) => v switch
    {
        <= 2 => "None / Minimal", <= 4 => "Low", <= 6 => "Moderate", <= 8 => "High", _ => "Heavy"
    };
    private static string GetGuardednessLabel(double v) => v switch
    {
        <= 2 => "Welcoming", <= 4 => "Friendly", <= 6 => "Watchful",
        <= 8 => "Protective", _ => "Very Guarded"
    };
    private static string GetAggressivenessLabel(double v) => v switch
    {
        <= 2 => "Very Gentle", <= 4 => "Calm", <= 6 => "Assertive",
        <= 8 => "Bold", _ => "Very Dominant"
    };
    private static string GetImmunityLabel(double v) => v switch
    {
        <= 2 => "Fragile", <= 4 => "Below Average", <= 6 => "Average",
        <= 8 => "Robust", _ => "Exceptional"
    };
    private static string GetLifespanLabel(double v) => v switch
    {
        <= 2 => "Short (< 7 yrs)", <= 4 => "Below Average", <= 6 => "Average (10–12 yrs)",
        <= 8 => "Long (13–15 yrs)", _ => "Very Long (16+ yrs)"
    };
    private static string GetGroomingLabel(double v) => v switch
    {
        <= 2 => "Minimal", <= 4 => "Occasional", <= 6 => "Regular",
        <= 8 => "Frequent", _ => "Salon-Level"
    };
    private static string GetVetVisitsLabel(double v) => v switch
    {
        <= 2 => "Rarely", <= 4 => "Occasional", <= 6 => "Routine",
        <= 8 => "Frequent", _ => "Very Frequent"
    };
    private static string GetLoyaltyLabel(double v) => v switch
    {
        <= 2 => "Independent", <= 4 => "Somewhat Independent", <= 6 => "Balanced",
        <= 8 => "Very Loyal", _ => "Devotedly Loyal"
    };
    private static string GetSingleOwnerLabel(double v) => v switch
    {
        <= 2 => "Great with Everyone", <= 4 => "Family Dog", <= 6 => "Adapts to Family",
        <= 8 => "Prefers One Person", _ => "One-Person Dog"
    };
}

File: BreedFinder/ViewModels/MainViewModel.swift
import Foundation
import Observation

// MARK: - Main view-model
@Observable
@MainActor
final class MainViewModel {

    // ── Sliders (0–10) — Living Situation ─────────────────────────────────────
    var sizeValue:           Double = 5
    var energyLevelValue:    Double = 5
    var aloneValue:          Double = 5   // alone tolerance

    // ── Sliders — Family & Social ─────────────────────────────────────────────
    var kidsValue:           Double = 5   // kids compatibility
    var petsValue:           Double = 5   // pet compatibility
    var singleOwnerValue:    Double = 5

    // ── Sliders — Personality ─────────────────────────────────────────────────
    var trainabilityValue:   Double = 5
    var barkingValue:        Double = 5   // barking level
    var guardednessValue:    Double = 5
    var aggressivenessValue: Double = 5
    var loyaltyValue:        Double = 5

    // ── Sliders — Care & Grooming ─────────────────────────────────────────────
    var sheddingValue:       Double = 5
    var groomingNeedsValue:  Double = 5

    // ── Sliders — Health & Lifespan ───────────────────────────────────────────
    var immunityValue:       Double = 5
    var lifespanValue:       Double = 5
    var vetVisitsValue:      Double = 5

    // ── Computed slider labels ────────────────────────────────────────────────
    var sizeLabel:           String { sizeText(sizeValue) }
    var energyLevelLabel:    String { energyText(energyLevelValue) }
    var aloneLabel:          String { aloneText(aloneValue) }
    var kidsLabel:           String { kidsText(kidsValue) }
    var petsLabel:           String { petsText(petsValue) }
    var singleOwnerLabel:    String { singleOwnerText(singleOwnerValue) }
    var trainabilityLabel:   String { trainabilityText(trainabilityValue) }
    var barkingLabel:        String { barkingText(barkingValue) }
    var guardednessLabel:    String { guardednessText(guardednessValue) }
    var aggressivenessLabel: String { aggressivenessText(aggressivenessValue) }
    var loyaltyLabel:        String { loyaltyText(loyaltyValue) }
    var sheddingLabel:       String { sheddingText(sheddingValue) }
    var groomingNeedsLabel:  String { groomingText(groomingNeedsValue) }
    var immunityLabel:       String { immunityText(immunityValue) }
    var lifespanLabel:       String { lifespanText(lifespanValue) }
    var vetVisitsLabel:      String { vetVisitsText(vetVisitsValue) }

    // ── Search ────────────────────────────────────────────────────────────────
    var searchText: String = ""

    var isSearching: Bool {
        !searchText.trimmingCharacters(in: .whitespaces).isEmpty
    }

    var searchResults: [DogBreed] {
        guard isSearching else { return [] }
        let q = searchText.lowercased().trimmingCharacters(in: .whitespaces)
        return BreedDatabase.all
            .filter { $0.name.lowercased().contains(q) }
            .sorted {
                // Exact prefix matches float to the top
                let aP = $0.name.lowercased().hasPrefix(q)
                let bP = $1.name.lowercased().hasPrefix(q)
                if aP != bP { return aP }
                return $0.name < $1.name
            }
    }

    var searchHeader: String {
        let q  = searchText.trimmingCharacters(in: .whitespaces)
        let n  = searchResults.count
        guard n > 0 else { return "No breeds match \u{201C}\(q)\u{201D}" }
        return "\(n) breed\(n == 1 ? "" : "s") matching \u{201C}\(q)\u{201D}"
    }

    // ── Results state ─────────────────────────────────────────────────────────
    var matchingBreeds:   [DogBreed] = []
    var filtersExpanded:  Bool = true
    var isResultsVisible: Bool = false
    var isLoading:        Bool = false
    var resultsHeader:    String = ""

    // ── Private ───────────────────────────────────────────────────────────────
    private let imageService = DogImageService()

    // ── Init ──────────────────────────────────────────────────────────────────
    init() {
        // Pre-warm the image cache for all breeds in the background
        let keys = BreedDatabase.all.map(\.dogCeoBreedKey)
        let svc  = imageService
        Task.detached(priority: .background) {
            await svc.prewarm(keys: keys)
        }
    }

    // ── Commands ──────────────────────────────────────────────────────────────

    func findBreeds() async {
        isLoading        = true
        isResultsVisible = false
        filtersExpanded  = false

        // Capture slider values before hopping off the main actor
        let (s, e, sh, g, ag, im, li, gr, v, lo, si) = (
            sizeValue, energyLevelValue, sheddingValue,
            guardednessValue, aggressivenessValue, immunityValue,
            lifespanValue, groomingNeedsValue, vetVisitsValue,
            loyaltyValue, singleOwnerValue
        )
        let (ki, pe, al, ba, tr) = (
            kidsValue, petsValue, aloneValue, barkingValue, trainabilityValue
        )

        // Run CPU-bound matching off the main thread
        let results = await Task.detached(priority: .userInitiated) {
            BreedDatabase.getMatchingBreeds(
                size: s, energy: e, shedding: sh,
                guardedness: g, aggressiveness: ag, immunity: im,
                lifespan: li, grooming: gr, vetVisits: v,
                loyalty: lo, singleOwner: si,
                kidsCompat: ki, petCompat: pe, aloneToler: al,
                barking: ba, trainability: tr
            )
        }.value

        matchingBreeds   = results
        resultsHeader    = "\(results.count) breeds found — sorted by best match"
        isLoading        = false
        isResultsVisible = true

        // Fetch images in the background; cards update as URLs arrive
        fetchImages(for: results)
    }

    func resetFilters() {
        sizeValue = 5; energyLevelValue = 5; aloneValue = 5
        kidsValue = 5; petsValue = 5; singleOwnerValue = 5
        trainabilityValue = 5; barkingValue = 5
        guardednessValue = 5; aggressivenessValue = 5; loyaltyValue = 5
        sheddingValue = 5; groomingNeedsValue = 5
        immunityValue = 5; lifespanValue = 5; vetVisitsValue = 5
    }

    // ── Background image fetching ─────────────────────────────────────────────

    private func fetchImages(for breeds: [DogBreed]) {
        let svc = imageService   // capture actor reference (sendable)
        Task.detached(priority: .utility) { [weak self] in
            // Process in batches of 4
            let chunks = stride(from: 0, to: breeds.count, by: 4).map {
                Array(breeds[$0..<min($0 + 4, breeds.count)])
            }
            for chunk in chunks {
                await withTaskGroup(of: (UUID, String?).self) { group in
                    for breed in chunk {
                        group.addTask {
                            let url = await svc.getImageUrl(for: breed.dogCeoBreedKey)
                            return (breed.id, url)
                        }
                    }
                    for await (breedId, url) in group {
                        guard let url else { continue }
                        await self?.applyImageUrl(breedId: breedId, url: url)
                    }
                }
            }
        }
    }

    // Called from the detached task; hops to main actor automatically
    // because MainViewModel is @MainActor.
    private func applyImageUrl(breedId: UUID, url: String) {
        if let idx = matchingBreeds.firstIndex(where: { $0.id == breedId }) {
            matchingBreeds[idx].imageUrl = url
        }
    }

    // ── Label helpers ─────────────────────────────────────────────────────────

    private func sizeText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Tiny"
        case ..<4:  return "Small"
        case ..<6:  return "Medium"
        case ..<8:  return "Large"
        default:    return "Giant"
        }
    }
    private func energyText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Couch Potato"
        case ..<4:  return "Low Energy"
        case ..<6:  return "Moderate"
        case ..<8:  return "High Energy"
        default:    return "Hyperactive"
        }
    }
    private func aloneText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Never Alone"
        case ..<4:  return "Needs Company"
        case ..<6:  return "Short Periods"
        case ..<8:  return "Independent"
        default:    return "Very Independent"
        }
    }
    private func kidsText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Not for Kids"
        case ..<4:  return "Adults Only"
        case ..<6:  return "Older Kids OK"
        case ..<8:  return "Good with Kids"
        default:    return "Great with Kids"
        }
    }
    private func petsText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Needs Solo Home"
        case ..<4:  return "Selective"
        case ..<6:  return "Can Adapt"
        case ..<8:  return "Friendly"
        default:    return "Loves All Pets"
        }
    }
    private func singleOwnerText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Great with Everyone"
        case ..<4:  return "Family Dog"
        case ..<6:  return "Adapts to Family"
        case ..<8:  return "Prefers One Person"
        default:    return "One-Person Dog"
        }
    }
    private func trainabilityText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Very Stubborn"
        case ..<4:  return "Challenging"
        case ..<6:  return "Average"
        case ..<8:  return "Quick Learner"
        default:    return "Exceptionally Easy"
        }
    }
    private func barkingText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Nearly Silent"
        case ..<4:  return "Quiet"
        case ..<6:  return "Moderate"
        case ..<8:  return "Vocal"
        default:    return "Very Vocal"
        }
    }
    private func sheddingText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "None / Minimal"
        case ..<4:  return "Low"
        case ..<6:  return "Moderate"
        case ..<8:  return "High"
        default:    return "Heavy"
        }
    }
    private func guardednessText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Welcoming"
        case ..<4:  return "Friendly"
        case ..<6:  return "Watchful"
        case ..<8:  return "Protective"
        default:    return "Very Guarded"
        }
    }
    private func aggressivenessText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Very Gentle"
        case ..<4:  return "Calm"
        case ..<6:  return "Assertive"
        case ..<8:  return "Bold"
        default:    return "Very Dominant"
        }
    }
    private func immunityText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Fragile"
        case ..<4:  return "Below Average"
        case ..<6:  return "Average"
        case ..<8:  return "Robust"
        default:    return "Exceptional"
        }
    }
    private func lifespanText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Short (< 7 yrs)"
        case ..<4:  return "Below Average"
        case ..<6:  return "Average (10–12 yrs)"
        case ..<8:  return "Long (13–15 yrs)"
        default:    return "Very Long (16+ yrs)"
        }
    }
    private func groomingText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Minimal"
        case ..<4:  return "Occasional"
        case ..<6:  return "Regular"
        case ..<8:  return "Frequent"
        default:    return "Salon-Level"
        }
    }
    private func vetVisitsText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Rarely"
        case ..<4:  return "Occasional"
        case ..<6:  return "Routine"
        case ..<8:  return "Frequent"
        default:    return "Very Frequent"
        }
    }
    private func loyaltyText(_ v: Double) -> String {
        switch v {
        case ..<2:  return "Independent"
        case ..<4:  return "Somewhat Independent"
        case ..<6:  return "Balanced"
        case ..<8:  return "Very Loyal"
        default:    return "Devotedly Loyal"
        }
    }
}
